### PR TITLE
Add transaction logic and tests to it.

### PR DESCRIPTION
**Pull request by bodikqlar**

### CODE REVIEW

File: app/models/transaction.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  #---------------------Associations----------------------
+  belongs_to :creditor, foreign_key: :creditor_id, class_name: 'User'
+  belongs_to :debitor, foreign_key: :debitor_id, class_name: 'User'
+  belongs_to :user
+  #---------------------Callbacks-------------------------
+  after_create { creditor.calculate_money_amount if creditor}
+  after_create { debitor.calculate_money_amount }
+  #---------------------Validations-----------------------
+  validates_presence_of :debitor_id
+  validates_presence_of :amount
```
Comment (maksymst) [inline app/models/transaction.rb]: this can be added to previous validation

File: app/models/transaction.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  validates_numericality_of :amount, greater_than: 0
+  validates_presence_of :user_id, if: 'creditor_id.nil?'
+  validates_presence_of :memo, if: 'creditor_id.nil?'
```
Comment (maksymst) [inline app/models/transaction.rb]: this can be added to previous validation

File: app/models/transaction.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  validate :can_afford?, unless: 'creditor_id.nil?'
+  validate :not_equal?, unless: 'creditor_id.nil?'
+
+  #---------------------Instance methods------------------
+  private
```
Comment (maksymst) [inline app/models/transaction.rb:16]: add new line below

File: app/models/transaction.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  def can_afford?
+    condition = creditor.calculate_money_amount - amount >= 0
+    errors.add(:date, "You don't have enough money.") unless condition
+  end
+
+  def not_equal?
+    errors.add(:date, 'Debitor and creditor are the same.') unless creditor.id != debitor_id
```
Comment (maksymst) [inline app/models/transaction.rb]: refactor condition

File: app/models/transaction.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  validates_presence_of :user_id, if: 'creditor_id.nil?'
+  validates_presence_of :memo, if: 'creditor_id.nil?'
+  validate :can_afford?, unless: 'creditor_id.nil?'
+  validate :not_equal?, unless: 'creditor_id.nil?'
+
+  #---------------------Instance methods------------------
+  private
+  def can_afford?
+    condition = creditor.calculate_money_amount - amount >= 0
+    errors.add(:date, "You don't have enough money.") unless condition
```
Comment (maksymst) [inline app/models/transaction.rb]: why do you add error to `date` field?

File: app/models/transaction.rb (hunk 4 above)
Comment (maksymst) [inline app/models/transaction.rb]: why do you add error to `date` field?

File: app/models/user.rb [hunk 6]
```diff
@@ -25,6 +28,34 @@ def full_name
     "#{first_name} #{last_name}"
   end
 
+  def calculate_money_amount
+    self.current_amount = debit_transactions.sum(:amount) - credit_transactions.sum(:amount)
+    save
```
Comment (maksymst) [inline app/models/user.rb]: why do you use save here?

File: app/models/user.rb [hunk 7]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    current_amount
+  end
+
+  def payable?
+    calculate_money_amount > 0
+  end
+
+  def increase_balance(**args)
```
Comment (maksymst) [inline app/models/user.rb]: use explicit argunments

File: app/models/user.rb [hunk 8]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    creditor_id = args[:creditor_id]
+    amount = args[:amount]
+    debit_transactions.create(creditor_id: creditor_id, amount: amount)
+  end
+
+  def decrease_balance(**args)
```
Comment (maksymst) [inline app/models/user.rb]: use explicit argunments

File: app/models/user.rb [hunk 9]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    debitor_id = args[:debitor_id]
+    amount = args[:amount]
+    credit_transactions.create(debitor_id: debitor_id, amount: amount)
+  end
+
+  def add_money(**args)
```
Comment (maksymst) [inline app/models/user.rb]: use explicit argunments

File: app/models/user.rb [hunk 10]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    save
+    current_amount
+  end
+
+  def payable?
+    calculate_money_amount > 0
+  end
+
+  def increase_balance(**args)
+    creditor_id = args[:creditor_id]
```
Comment (maksymst) [inline app/models/user.rb]: get rid reassigning

File: app/models/user.rb [hunk 11]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    amount = args[:amount]
+    debit_transactions.create(creditor_id: creditor_id, amount: amount)
+  end
+
+  def decrease_balance(**args)
+    debitor_id = args[:debitor_id]
```
Comment (maksymst) [inline app/models/user.rb]: get rid reassigning

File: app/models/user.rb [hunk 12]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    amount = args[:amount]
```
Comment (MaximSt) [inline app/models/user.rb]: get rid reassigning

File: app/models/user.rb [hunk 13]
```diff
@@ -25,6 +28,34 @@ def full_name
...
+    credit_transactions.create(debitor_id: debitor_id, amount: amount)
+  end
+
+  def add_money(**args)
+    debitor_id = args[:debitor_id]
```
Comment (maksymst) [inline app/models/user.rb]: get rid reassigning

File: spec/factories/transaction.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :transaction, class: 'Transaction' do
+    amount { rand(1..1000) }
```
Comment (maksymst) [inline spec/factories/transaction.rb:5]: factory is invalid

File: spec/models/transaction_spec.rb [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    it { is_expected.to validate_presence_of(:debitor_id) }
+    it { is_expected.to validate_presence_of(:amount) }
+    it { is_expected.to validate_numericality_of(:amount).is_greater_than(0) }
+
+    context 'if creditor is present' do
+      it { expect(transaction_without_creditor).to be_valid }
+    end
+
+    context 'if creditor is not present' do
+      before { allow(subject).to receive(:creditor) { nil } }
```
Comment (maksymst) [inline spec/models/transaction_spec.rb:36]: add new line below

File: spec/models/user_spec.rb [hunk 16]
```diff
@@ -18,4 +21,34 @@
...
   end
+
+  describe 'payment methods' do
+    let!(:debitor) { create(:user) }
+    let!(:creditor) { create(:user) }
+    let!(:admin) { create(:user) }
+    let(:transaction) { build :transaction }
+    let(:last_transaction) { Transaction.last }
+
+    before(:each) { admin.add_money(debitor_id: creditor.id, amount: transaction.amount, memo: transaction.memo) }
```
Comment (maksymst) [inline spec/models/user_spec.rb]: add new line below

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -18,4 +21,34 @@
...
+    it 'increases money in debitor' do
+      expect(debitor.increase_balance(creditor_id: creditor.id, amount: transaction.amount)).to eq(last_transaction)
+    end
+
+    it 'decreases money amount in creditor' do
+      expect(creditor.decrease_balance(debitor_id: debitor.id, amount: transaction.amount)).to eq(last_transaction)
+    end
+
+    it 'checks payable ability of user' do
+      expect(debitor.payable?).to be false
```
Comment (maksymst) [inline spec/models/user_spec.rb]: rspec 3 has specific method for this

File: spec/models/user_spec.rb [hunk 18]
```diff
@@ -18,4 +21,34 @@
...
+      expect(creditor.payable?).to be true
```
Comment (maksymst) [inline spec/models/user_spec.rb]: rspec 3 has specific method for this

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -18,4 +21,34 @@
...
+
+    it 'increases money in debitor' do
+      expect(debitor.increase_balance(creditor_id: creditor.id, amount: transaction.amount)).to eq(last_transaction)
+    end
+
+    it 'decreases money amount in creditor' do
+      expect(creditor.decrease_balance(debitor_id: debitor.id, amount: transaction.amount)).to eq(last_transaction)
+    end
+
+    it 'checks payable ability of user' do
```
Comment (maksymst) [inline spec/models/user_spec.rb]: there are no conditions here.

File: spec/models/user_spec.rb [hunk 20]
```diff
@@ -18,4 +21,34 @@
...
+  describe 'payment methods' do
+    let!(:debitor) { create(:user) }
+    let!(:creditor) { create(:user) }
+    let!(:admin) { create(:user) }
+    let(:transaction) { build :transaction }
+    let(:last_transaction) { Transaction.last }
+
+    before(:each) { admin.add_money(debitor_id: creditor.id, amount: transaction.amount, memo: transaction.memo) }
+    it 'calculates balance' do
+      expect(creditor.calculate_money_amount).to eq(transaction.amount)
```
Comment (maksymst) [inline spec/models/user_spec.rb]: refactor this

File: spec/models/user_spec.rb [hunk 21]
```diff
@@ -18,4 +21,34 @@
...
+    end
+
+    it 'adds money to debitor' do
+      expect(last_transaction.amount).to eq(transaction.amount)
```
Comment (maksymst) [inline spec/models/user_spec.rb]: refactor this

File: spec/models/user_spec.rb [hunk 22]
```diff
@@ -18,4 +21,34 @@
       expect(build(:user)).to be_valid
     end
   end
+
+  describe 'payment methods' do
```
Comment (maksymst) [inline spec/models/user_spec.rb]: use describe for each method

File: spec/models/user_spec.rb [hunk 23]
```diff
@@ -18,4 +21,34 @@
...
+    it 'calculates balance' do
+      expect(creditor.calculate_money_amount).to eq(transaction.amount)
+    end
+
+    it 'adds money to debitor' do
+      expect(last_transaction.amount).to eq(transaction.amount)
+    end
+
+    it 'increases money in debitor' do
+      expect(debitor.increase_balance(creditor_id: creditor.id, amount: transaction.amount)).to eq(last_transaction)
```
Comment (maksymst) [inline spec/models/user_spec.rb]: refactor this

File: spec/models/user_spec.rb [hunk 24]
```diff
@@ -18,4 +21,34 @@
...
+    end
+
+    it 'decreases money amount in creditor' do
+      expect(creditor.decrease_balance(debitor_id: debitor.id, amount: transaction.amount)).to eq(last_transaction)
```
Comment (maksymst) [inline spec/models/user_spec.rb]: refactor this

File: spec/models/transaction_spec.rb [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+
+    it { is_expected.to validate_presence_of(:debitor_id) }
+    it { is_expected.to validate_presence_of(:amount) }
+    it { is_expected.to validate_numericality_of(:amount).is_greater_than(0) }
+
+    context 'if creditor is present' do
+      let(:transaction_with_sponsor) do
+        admin.add_money(sponsor.id, 100, 'Add some money')
+        build(:transaction, creditor: sponsor, debitor: recipient, amount: sponsor.calculate_money_amount)
+      end
```
Comment (maksymst) [inline spec/models/transaction_spec.rb:28]: add new line below

File: spec/models/transaction_spec.rb [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+      it { expect(transaction_with_sponsor).to be_valid}
```
Comment (maksymst) [inline spec/models/transaction_spec.rb]: missing space

File: spec/models/transaction_spec.rb [hunk 27]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+RSpec.describe Transaction, type: :model do
+
+  describe 'association' do
+    it { is_expected.to belong_to(:user)}
```
Comment (maksymst) [inline spec/models/transaction_spec.rb]: missing space

File: spec/models/transaction_spec.rb [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+    it { is_expected.to belong_to(:debitor)}
```
Comment (maksymst) [inline spec/models/transaction_spec.rb]: missing space

File: spec/models/transaction_spec.rb [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+    it { is_expected.to belong_to(:creditor)}
```
Comment (maksymst) [inline spec/models/transaction_spec.rb]: missing space

File: spec/models/transaction_spec.rb [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+    context 'if creditor is present' do
+      let(:transaction_with_sponsor) do
+        admin.add_money(sponsor.id, 100, 'Add some money')
+        build(:transaction, creditor: sponsor, debitor: recipient, amount: sponsor.calculate_money_amount)
+      end
+      it { expect(transaction_with_sponsor).to be_valid}
+    end
+
+    context 'if creditor is not present' do
+      let(:invalid_transaction) { build(:transaction, creditor: admin, debitor: admin) }
```
Comment (maksymst) [inline spec/models/transaction_spec.rb:34]: add new line below

File: spec/models/transaction_spec.rb [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+      before { allow(subject).to receive(:creditor) { nil } }
+
+      it { is_expected.to validate_presence_of(:user_id) }
+      it { is_expected.to validate_presence_of(:memo) }
+      it { expect(invalid_transaction).not_to be_valid}
```
Comment (maksymst) [inline spec/models/transaction_spec.rb]: missing space

File: spec/models/user_spec.rb [hunk 32]
```diff
@@ -18,4 +21,39 @@
       expect(build(:user)).to be_valid
     end
   end
+
+  let(:recipient) { create(:user) }
+  let(:sponsor) { create(:user) }
+  let(:admin) { create(:user) }
+  let(:money_amount) { 20 }
+  let(:memo) {  Faker::Company.catch_phrase }
```
Comment (MaximSt) [inline spec/models/user_spec.rb]: company name for this porpose is confusing

File: spec/models/user_spec.rb [hunk 33]
```diff
@@ -18,4 +21,39 @@
...
+
+  describe '.add_money' do
+    it { expect{ admin.add_money(sponsor.id, money_amount, memo) }.to change{ sponsor.calculate_money_amount }.by(money_amount) }
+  end
+
+  describe '.increase_balance' do
```
Comment (maksymst) [inline spec/models/user_spec.rb]: old description

File: spec/models/user_spec.rb [hunk 34]
```diff
@@ -18,4 +21,39 @@
...
+    it 'checks money increasing' do
```
Comment (maksymst) [inline spec/models/user_spec.rb]: description is confusing

File: spec/models/user_spec.rb [hunk 35]
```diff
@@ -18,4 +21,39 @@
...
+      admin.add_money(sponsor.id, money_amount, memo)
+      expect{ sponsor.transfer_money(recipient.id, money_amount) }.to change{ recipient.calculate_money_amount }.by(money_amount)
+    end
+  end
+
+  describe '.calculate_money_amount' do
+    it 'checks money amount' do
```
Comment (maksymst) [inline spec/models/user_spec.rb]: description is confusing

File: spec/models/user_spec.rb [hunk 36]
```diff
@@ -18,4 +21,39 @@
...
+      admin.add_money(sponsor.id, money_amount, memo)
+      expect(sponsor.calculate_money_amount).to eq(money_amount)
+    end
+  end
+
+  describe '#payable?' do
+    before { admin.add_money(sponsor.id, money_amount, memo) }
```
Comment (maksymst) [inline spec/models/user_spec.rb:50]: add new line below

File: spec/models/user_spec.rb [hunk 37]
```diff
@@ -18,4 +21,39 @@
...
+    context "doesn't have money" do
+      it { expect(recipient.payable?).to be_falsey }
+    end
```
Comment (maksymst) [inline spec/models/user_spec.rb:54]: add new line below

File: app/models/user.rb [hunk 38]
```diff
@@ -15,6 +15,9 @@ class User < ActiveRecord::Base
   has_many :movies, dependent: :destroy
   has_one :address, dependent: :destroy
 
+  has_many :debit_transactions, foreign_key: :debitor_id, class_name: 'Transaction'
```
Comment (maksymst) [inline app/models/user.rb]: `dependent` is missing

File: app/models/user.rb [hunk 39]
```diff
@@ -15,6 +15,9 @@ class User < ActiveRecord::Base
...
+  has_many :credit_transactions, foreign_key: :creditor_id, class_name: 'Transaction'
```
Comment (maksymst) [inline app/models/user.rb]: `dependent` is missing

File: app/models/user.rb [hunk 40]
```diff
@@ -15,6 +15,9 @@ class User < ActiveRecord::Base
...
+  has_many :admin_transactions, foreign_key: :user_id, class_name: 'Transaction'
```
Comment (maksymst) [inline app/models/user.rb]: `dependent` is missing

**Pull request merged**